packages/payment-processor/src/payment/erc20.ts

```diff
@@ -55,18 +66,70 @@ export async function hasErc20Approval(
   account: string,
   provider: Provider = getNetworkProvider(request),
 ): Promise<boolean> {
-  const erc20Contract = ERC20Contract.connect(request.currencyInfo.value, provider);
-  const allowance = await erc20Contract.allowance(account, getProxyAddress(request));
-  return allowance.gt(request.expectedAmount);
+  return checkErc20Allowance(
+    account,
+    getProxyAddress(request),
+    provider,
+    request.currencyInfo,
+    new BigNumber(request.expectedAmount)
+  )
+}
+
+/**
+ * Checks if a spender has enough allowance from an ERC20 token owner to pay an amount.
+ * @param ownerAddress address of the owner
+ * @param spenderAddress address of the spender
+ * @param provider the web3 provider. Defaults to Etherscan.
+ * @param paymentCurrency ERC20 currency
+ * @param amount
+ */
+export async function checkErc20Allowance(
+  ownerAddress: string,
+  spenderAddress: string,
+  provider: Provider,
+  paymentCurrency: RequestLogicTypes.ICurrency,
+  amount: BigNumber,
+): Promise<boolean> {
+  if (paymentCurrency.type !== RequestLogicTypes.CURRENCY.ERC20) {
+    throw new Error('Trying to check the allowance of a non-ERC20 currency');
+  }
+  const erc20Contract = ERC20Contract.connect(paymentCurrency.value, provider);
+  const allowance = await erc20Contract.allowance(ownerAddress, spenderAddress);
+  return allowance.gte(amount);
 }
 
 /**
  * Processes the approval transaction of the targeted ERC20.
  * @param request request to pay
  * @param provider the web3 provider. Defaults to Etherscan.
  * @param overrides optionally, override default transaction values, like gas.
- * @param paymentNetworkId the payment network id
- * @param proxyContractAddress the address of the proxy contract to set the approval.
+ */
+export async function approveErc20IfNeeded(
+  request: ClientTypes.IRequestData,
+  account: string,
+  provider: Provider = getNetworkProvider(request),
+  overrides?: ITransactionOverrides,
+): Promise<ContractTransaction | void> {
+  if (!hasErc20Approval(request, account, provider)) {
+    const signer = getSigner(provider);
+    const encodedTx = encodeApproveErc20(request, signer);
+    const tokenAddress = request.currencyInfo.value;
+    const tx = await signer.sendTransaction({
+      data: encodedTx,
+      to: tokenAddress,
+      value: 0,
+      ...overrides,
+    });
+    return tx;
```

Comment:
this seems exactly the same as `approveErc20`?

Reply:
Oopsie 👍  Thanks